### PR TITLE
[CUSOLVER] Add generic routines

### PR DESCRIPTION
**Pull request by amontoison**
Interface the new generic routines of CUSOLVER.
`systrs` and `trtri` were not available before.

### CODE REVIEW

**Comment by amontoison** [comment]
@maleadt 
`CUSOLVER` added generic routines for LAPACK functions and started to deprecate the legacy routines.
They are doing something similar to `CUSPARSE`.

The new main routines that I added are `sytrs` and `trtri`.
We will be able to add specialized methods for inverting dense triangular matrices and solving dense symmetric / hermitian linear systems.

Except `gesvdp`, `sysevdx` and `gesvdr`, I added tests for all new routines.
I will uncomment `gesvdp`, `sysevdx` and `gesvdr` when I will add the tests but I propose to do that in a different PR.

File: lib/cusolver/libcusolver.jl [hunk 1]
```diff
@@ -1610,7 +1610,7 @@ end
     @ccall libcusolver.cusolverDnXtrtri_bufferSize(handle::cusolverDnHandle_t,
                                                    uplo::cublasFillMode_t,
                                                    diag::cublasDiagType_t, n::Int64,
-                                                   dataTypeA::cudaDataType, A::Ptr{Cvoid},
+                                                   dataTypeA::cudaDataType, A::CuPtr{Cvoid},
```
Comment (maleadt) [inline lib/cusolver/libcusolver.jl:1613]: Can you update these changes in https://github.com/JuliaGPU/CUDA.jl/blob/master/res/wrap/cusolver.toml?

File: lib/utils/call.jl [hunk 2]
```diff
@@ -52,7 +52,7 @@ This helper protects against the rare but real issue of the workspace size gette
 different results based on the GPU device memory pressure, which might change _after_
 initial allocation of the workspace (which can cause a GC collection).
 """
-@inline with_workspace(f, size::Union{Integer,Function}, fallback::Union{Nothing,Integer}=nothing; keep::Bool=false) =
+@inline with_workspace(f::Function, size::Union{Integer,Function}, fallback::Union{Nothing,Integer}=nothing; keep::Bool=false) =
```
Comment (maleadt) [inline lib/utils/call.jl]: `Base.Callable`?

**Review by maleadt — APPROVED** [review]
Some minor nits, but LGTM. Thanks!

File: lib/cusolver/libcusolver.jl (hunk 1 above)
Comment (amontoison) [reply lib/cusolver/libcusolver.jl:1613]: Tim, the routines with the wrong argument types are not in `res/wrap/cusolver.toml`.
How do you (re)generate this toml file?

Comment (maleadt) [reply lib/cusolver/libcusolver.jl:1613]: `cusolver.tomlm` is not generated, it's a human-curated list of definitions, and the changes that need to be made to the argument types. When we parse the C headers, we can only see pointers, and not the difference between CPU and GPU pointers, so we cannot autogenerate this file.

Comment (amontoison) [reply lib/cusolver/libcusolver.jl:1613]: Thanks :+1: 
I updated `cusolver.toml`.

**Pull request merged**